homeassistant/components/devolo_home_control/__init__.py

```diff
@@ -66,31 +88,21 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
 
     await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
 
-    def shutdown(event: Event) -> None:
-        for gateway in hass.data[DOMAIN][entry.entry_id]["gateways"]:
-            gateway.websocket_disconnect(
-                f"websocket disconnect requested by {EVENT_HOMEASSISTANT_STOP}"
-            )
-
-    # Listen when EVENT_HOMEASSISTANT_STOP is fired
-    hass.data[DOMAIN][entry.entry_id]["listener"] = hass.bus.async_listen_once(
-        EVENT_HOMEASSISTANT_STOP, shutdown
-    )
-
     return True
 
 
-async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+async def async_unload_entry(
+    hass: HomeAssistant, entry: DevoloHomeControlConfigEntry
+) -> bool:
     """Unload a config entry."""
     unload = await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
     await asyncio.gather(
         *(
             hass.async_add_executor_job(gateway.websocket_disconnect)
-            for gateway in hass.data[DOMAIN][entry.entry_id]["gateways"]
+            for gateway in entry.runtime_data.gateways
         )
     )
-    hass.data[DOMAIN][entry.entry_id]["listener"]()
-    hass.data[DOMAIN].pop(entry.entry_id)
+    entry.runtime_data.listener()
```

Comment:
```suggestion
```
With `entry.async_on_unload`, this can now be removed.

Thinking a bit further, it might make sense to remove the `DevoloHomeControlData` type completely as it only stores `list[HomeControl]`.